trieste/models/gpflow/architectures.py

```diff
@@ -0,0 +1,223 @@
+
+    gpflow.set_trainable(model.likelihood, (not noise_free))
+    if classification and noise_free:
+        gpflow.set_trainable(model.kernel.variance, False)
+
+    return model
+
+
+def build_svgp(
+    data: Dataset,
+    search_space: SearchSpace,
+    classification: bool = False,
+    kernel_priors: bool = True,
+    trainable_likelihood: bool = False,
+    likelihood_variance: Optional[float] = None,
+    num_inducing_points: Optional[int] = None,
+    trainable_inducing_points: bool = False,
+) -> SVGP:
+    empirical_mean, empirical_variance, num_data_points = _get_data_stats(data)
+
+    if classification:
+        empirical_variance = 1.0
+        empirical_mean = 0.0
+        model_likelihood = gpflow.likelihoods.Bernoulli()
+    else:
+        if likelihood_variance is None:
+            noise_variance = 0.1 * empirical_variance
+        else:
+            noise_variance = likelihood_variance
+        model_likelihood = gpflow.likelihoods.Gaussian(noise_variance)
+
+    kernel = _get_kernel(
+        empirical_variance, search_space, kernel_priors, classification, False
+    )
+    mean = gpflow.mean_functions.Constant(empirical_mean)
+
+    inducing_points = _get_inducing_points(search_space, num_inducing_points)
+
+    model = SVGP(
+        kernel,
+        model_likelihood,
+        inducing_points,
+        mean_function=mean,
+        num_data=num_data_points,
+    )
+
+    gpflow.set_trainable(model.likelihood, trainable_likelihood)
+    gpflow.set_trainable(model.inducing_variable, trainable_inducing_points)
+
+    return model
+
+
+def _get_data_stats(data: Dataset) -> tuple[TensorType, TensorType, int]:
+    empirical_variance = tf.math.reduce_variance(data.observations)
+    empirical_mean = tf.math.reduce_mean(data.observations)
+    num_data_points = len(data.observations)
+
+    return empirical_mean, empirical_variance, num_data_points
+
+
+def _get_kernel(
+    variance: TensorType,
+    search_space: SearchSpace,
+    priors: bool,
+    classification: bool = False,
+    noise_free: bool = False,
+) -> gpflow.kernels.Kernel:
+    
+    lengthscales = 0.2 * (search_space.upper - search_space.lower) * math.sqrt(search_space.dimension)
+    kernel = gpflow.kernels.Matern52(variance=variance, lengthscales=lengthscales)
+
+    if priors:
+        prior_scale = tf.cast(1.0, dtype=gpflow.default_float())
+        if not (classification and noise_free):
+            kernel.variance.prior = tfp.distributions.LogNormal(tf.math.log(variance), prior_scale)
+        kernel.lengthscales.prior = tfp.distributions.LogNormal(
+            tf.math.log(kernel.lengthscales), prior_scale
+        )
+
+    return kernel
+
+
+def _get_inducing_points(search_space: SearchSpace, num_inducing_points: Optional[int]) -> TensorType:
+    if num_inducing_points is None:
+        num_inducing_points = min(500, 25 * search_space.upper.shape[-1])
+
+    inducing_points = search_space.sample_sobol(num_inducing_points)
```

Comment:
I've never tried using sobol samples as inducing points, but doesn't it have the risk that if they are not trainable your BO loop will become stuck (it can't reduce variance where its selecting the latest data)? It seems suboptimal on small datasets on which BO is usually run, in which case you may have enough inducing points to put them on the data exactly?

Reply:
if you check @vpicheny comments above, to avoid retracing we set inducing points here and use them throughout the BO loop, so we would want to set them to something larger than the dataset - though the risk of getting stuck didn't occur to me...

Reply:
To avoid getting stuck we need to control inducing points as the optimization progresses. This is future functionality (e.g. what we did for the S-GP-TS paper)

Reply:
in discrete and product search spaces we cannot use sobol sampling, we switch to the usual random sampling in those cases?